File: analysis/sweep/collect_sweep_results.py
"""Collect MLM sweep results into a CSV.

Usage:
    python scripts/collect_sweep_results.py
    python scripts/collect_sweep_results.py --sweep runs/other_sweep --out results/other.csv
"""

import argparse
import csv
import json
import re
import sys
from pathlib import Path

SWEEP_DIR = Path("runs/chembl36_small_mask_mlm_lr_sweep")
OUT_FILE = SWEEP_DIR / "sweep_results.csv"

# Columns pulled from all_results.json (in order). These legacy names are kept
# for backwards compatibility with existing notebooks. The final_eval_* columns
# below make the source explicit for new analysis.
METRIC_KEYS = [
    "eval_loss",
    "eval_masked_accuracy",
    "eval_perplexity",
    "train_loss",
    "epoch",
    "num_parameters",
    "train_runtime",
    "eval_runtime",
]

FINAL_EVAL_KEYS = [
    "eval_loss",
    "eval_masked_accuracy",
    "eval_perplexity",
]

# Regex to parse dir names like: mask_hetero_span__mlm_0p15__lr_1e-4
_DIR_RE = re.compile(r"mask_(?P<strategy>.+?)__mlm_(?P<mlm_prob>[0-9p]+)__lr_(?P<lr>[0-9e\-+.]+)$")


def _parse_mlm_prob(raw: str) -> float:
    return float(raw.replace("p", "."))


def _parse_dir_name(name: str) -> dict | None:
    m = _DIR_RE.match(name)
    if not m:
        return None
    return {
        "strategy": m.group("strategy"),
        "mlm_prob": _parse_mlm_prob(m.group("mlm_prob")),
        "lr_from_name": m.group("lr"),
    }


def _load_json(path: Path) -> dict:
    return json.loads(path.read_text()) if path.exists() else {}


def _numeric(value: object) -> float | None:
    if isinstance(value, bool):
        return None
    if isinstance(value, (int, float)):
        return float(value)
    return None


def _eval_history(trainer_state: dict) -> list[dict]:
    history = trainer_state.get("log_history", [])
    if not isinstance(history, list):
        return []
    return [
        entry
        for entry in history
        if isinstance(entry, dict) and _numeric(entry.get("eval_loss")) is not None
    ]


def _best_logged_eval(history: list[dict]) -> dict:
    if not history:
        return {}
    return min(history, key=lambda entry: float(entry["eval_loss"]))


def _last_logged_eval(history: list[dict]) -> dict:
    return history[-1] if history else {}


def _checkpoint_step(checkpoint: object) -> int | str:
    if not isinstance(checkpoint, str) or "-" not in checkpoint:
        return ""
    raw_step = checkpoint.rsplit("-", 1)[-1]
    try:
        return int(raw_step)
    except ValueError:
        return raw_step


def collect(sweep_dir: Path) -> list[dict]:
    rows = []
    for run_dir in sorted(sweep_dir.iterdir()):
        if not run_dir.is_dir():
            continue
        parsed = _parse_dir_name(run_dir.name)
        if parsed is None:
            continue

        results_path = run_dir / "all_results.json"
        args_path = run_dir / "run_args.json"
        if not results_path.exists():
            print(f"  skip {run_dir.name}: no all_results.json", file=sys.stderr)
            continue

        results = _load_json(results_path)
        args = _load_json(args_path)
        trainer_state = _load_json(run_dir / "trainer_state.json")
        eval_history = _eval_history(trainer_state)
        best_logged = _best_logged_eval(eval_history)
        last_logged = _last_logged_eval(eval_history)
        best_checkpoint = trainer_state.get("best_model_checkpoint", "")
        load_best_model_at_end = args.get("load_best_model_at_end", "")
        metric_source = (
            "best_model_final_eval"
            if load_best_model_at_end is not False and best_checkpoint
            else "final_model_eval"
        )

        row = {
            "run": run_dir.name,
            "size": args.get("model_size", ""),
            "strategy": parsed["strategy"],
            "mlm_prob": parsed["mlm_prob"],
            "learning_rate": args.get("learning_rate", parsed["lr_from_name"]),
            "eval_masking_strategy": args.get("masking_strategy", parsed["strategy"]),
            "eval_mlm_probability": args.get("mlm_probability", parsed["mlm_prob"]),
            "load_best_model_at_end": load_best_model_at_end,
            "metric_source": metric_source,
            "metric_note": (
                "Final eval after training; if load_best_model_at_end was enabled, "
                "the best checkpoint was loaded first. Validation masking uses this "
                "run-specific masking_strategy and mlm_probability."
            ),
            "best_checkpoint": best_checkpoint,
            "best_step": trainer_state.get("best_global_step", "")
            or _checkpoint_step(best_checkpoint),
            "best_logged_eval_loss": best_logged.get("eval_loss", ""),
            "best_logged_eval_masked_accuracy": best_logged.get("eval_masked_accuracy", ""),
            "last_logged_eval_step": last_logged.get("step", ""),
            "last_logged_eval_loss": last_logged.get("eval_loss", ""),
            "last_logged_eval_masked_accuracy": last_logged.get("eval_masked_accuracy", ""),
        }
        for key in FINAL_EVAL_KEYS:
            row[f"final_{key}"] = results.get(key, "")
        for key in METRIC_KEYS:
            row[key] = results.get(key, "")

        rows.append(row)
    return rows


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--sweep", default=str(SWEEP_DIR))
    parser.add_argument("--out", default=None)
    args = parser.parse_args()

    sweep_dir = Path(args.sweep)
    out_file = Path(args.out) if args.out else sweep_dir / "sweep_results.csv"

    rows = collect(sweep_dir)
    if not rows:
        print("No results found.", file=sys.stderr)
        sys.exit(1)

    fieldnames = list(rows[0].keys())
    with out_file.open("w", newline="") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(rows)

    print(f"Wrote {len(rows)} rows to {out_file}")


if __name__ == "__main__":
    main()

File: R/collect_sweep_results.R
#!/usr/bin/env Rscript
# Collect MLM sweep results into CSV(s).
#
# Outputs:
#   <sweep>/sweep_results.csv          — one row per run, all training metrics
#   <sweep>/fixed_eval_collected.csv   — fixed-eval per-prob results joined with
#                                        sweep metadata (written only if source exists)
#
# Usage:
#   Rscript R/collect_sweep_results.R
#   Rscript R/collect_sweep_results.R --sweep runs/other_sweep --out results/other.csv
suppressPackageStartupMessages({
  library(tidyverse)
  library(jsonlite)
  library(fs)
  library(stringr)
  library(argparse)
})

# =============================================================================
# Configuration
# =============================================================================
DEFAULT_SWEEP_DIR <- "runs/chembl36_small_mask_mlm_lr_sweep"

DIR_RE <- "^mask_(.+?)__mlm_([0-9p]+)__lr_([0-9e\\-+.]+)$"

FINAL_EVAL_KEYS <- c("eval_loss", "eval_masked_accuracy", "eval_perplexity")

TRAINING_METRIC_KEYS <- c(
  "epoch",
  "train_loss",
  "num_parameters",
  "total_flos",
  "train_runtime",
  "train_samples_per_second",
  "train_steps_per_second",
  "train_samples_streaming",
  "eval_runtime",
  "eval_samples_per_second",
  "eval_steps_per_second"
)

RUN_ARG_KEYS <- c(
  "max_steps",
  "warmup_steps",
  "weight_decay",
  "per_device_train_batch_size",
  "gradient_accumulation_steps",
  "eval_steps",
  "span_p_geom",
  "span_max_length",
  "heteroatom_start_weight",
  "max_seq_length",
  "seed"
)

# =============================================================================
# Helpers
# =============================================================================
parse_mlm_prob <- function(raw) {
  as.numeric(str_replace_all(raw, "p", "."))
}

parse_dir_name <- function(name) {
  m <- str_match(name, DIR_RE)
  if (is.na(m[1, 1])) return(NULL)
  tibble(
    strategy    = m[1, 2],
    mlm_prob    = parse_mlm_prob(m[1, 3]),
    lr_from_name = m[1, 4]
  )
}

load_json_safe <- function(path) {
  if (!file_exists(path)) return(list())
  fromJSON(path, simplifyVector = FALSE)
}

as_scalar <- function(x, default = NA_character_) {
  if (is.null(x) || length(x) == 0) return(default)
  x[[1]]
}

get_chr <- function(x, key, default = NA_character_) {
  val <- x[[key]]
  if (is.null(val)) return(default)
  as.character(as_scalar(val, default))
}

get_dbl <- function(x, key, default = NA_real_) {
  val <- x[[key]]
  if (is.null(val) || is.logical(val)) return(default)
  parsed <- suppressWarnings(as.numeric(as_scalar(val, NA_character_)))
  if (is.na(parsed)) default else parsed
}

# =============================================================================
# Trainer state helpers
# =============================================================================
eval_history <- function(trainer_state) {
  history <- trainer_state[["log_history"]]
  if (is.null(history) || !is.list(history)) return(tibble())
  tibble(entry = history) |>
    mutate(
      eval_loss = map_dbl(entry, ~ {
        val <- .x[["eval_loss"]]
        if (is.numeric(val) && length(val) == 1 && !is.na(val)) as.numeric(val) else NA_real_
      })
    ) |>
    filter(!is.na(eval_loss))
}

best_logged_eval <- function(history_tbl) {
  if (nrow(history_tbl) == 0) return(list())
  history_tbl |> slice_min(eval_loss, n = 1, with_ties = FALSE) |> pull(entry) |> pluck(1)
}

last_logged_eval <- function(history_tbl) {
  if (nrow(history_tbl) == 0) return(list())
  history_tbl |> slice_tail(n = 1) |> pull(entry) |> pluck(1)
}

checkpoint_step <- function(checkpoint) {
  if (is.null(checkpoint) || !nzchar(checkpoint) || !str_detect(checkpoint, "-")) return(NA_integer_)
  parts <- str_split(checkpoint, "-", simplify = TRUE)
  parsed <- suppressWarnings(as.integer(parts[1, ncol(parts)]))
  if (!is.na(parsed)) parsed else NA_integer_
}

# =============================================================================
# Best-run metadata
# =============================================================================
load_best_run_names <- function(sweep_dir) {
  span     <- load_json_safe(path(sweep_dir, "best_span_run.json"))
  standard <- load_json_safe(path(sweep_dir, "best_standard_run.json"))
  list(
    span     = get_chr(span,     "run_name", NA_character_),
    standard = get_chr(standard, "run_name", NA_character_)
  )
}

# =============================================================================
# Per-run collection
# =============================================================================
collect_one_run <- function(run_dir) {
  run_name <- path_file(run_dir)
  parsed   <- parse_dir_name(run_name)
  if (is.null(parsed)) return(NULL)

  results_path <- path(run_dir, "all_results.json")
  if (!file_exists(results_path)) {
    message("  skip ", run_name, ": no all_results.json")
    return(NULL)
  }

  results      <- load_json_safe(results_path)
  args         <- load_json_safe(path(run_dir, "run_args.json"))
  trainer      <- load_json_safe(path(run_dir, "trainer_state.json"))
  history_tbl  <- eval_history(trainer)
  best_logged  <- best_logged_eval(history_tbl)
  last_logged  <- last_logged_eval(history_tbl)

  best_checkpoint       <- get_chr(trainer, "best_model_checkpoint", "")
  load_best_at_end      <- get_chr(args, "load_best_model_at_end", "")
  metric_source <- if (!identical(load_best_at_end, "FALSE") && nzchar(best_checkpoint)) {
    "best_model_final_eval"
  } else {
    "final_model_eval"
  }

  best_step <- get_dbl(trainer, "best_global_step", NA_real_)
  if (is.na(best_step)) best_step <- checkpoint_step(best_checkpoint)

  identity <- tibble(
    run              = run_name,
    size             = get_chr(args, "model_size", NA_character_),
    strategy         = parsed$strategy,
    mlm_prob         = parsed$mlm_prob,
    learning_rate    = get_dbl(args, "learning_rate", as.numeric(parsed$lr_from_name)),
    masking_strategy = get_chr(args, "masking_strategy", parsed$strategy),
    mlm_probability  = get_dbl(args, "mlm_probability",  parsed$mlm_prob)
  )

  checkpoint_info <- tibble(
    load_best_model_at_end = load_best_at_end,
    metric_source          = metric_source,
    best_checkpoint        = best_checkpoint,
    best_step              = best_step
  )

  logged <- tibble(
    best_logged_eval_loss             = get_dbl(best_logged, "eval_loss"),
    best_logged_eval_masked_accuracy  = get_dbl(best_logged, "eval_masked_accuracy"),
    last_logged_eval_step             = get_dbl(last_logged, "step"),
    last_logged_eval_loss             = get_dbl(last_logged, "eval_loss"),
    last_logged_eval_masked_accuracy  = get_dbl(last_logged, "eval_masked_accuracy")
  )

  final_metrics <- FINAL_EVAL_KEYS |>
    set_names(paste0("final_", FINAL_EVAL_KEYS)) |>
    map(~ get_dbl(results, .x)) |>
    as_tibble()

  training_metrics <- TRAINING_METRIC_KEYS |>
    map(~ get_dbl(results, .x)) |>
    set_names(TRAINING_METRIC_KEYS) |>
    as_tibble()

  run_args <- RUN_ARG_KEYS |>
    map(~ get_dbl(args, .x)) |>
    set_names(RUN_ARG_KEYS) |>
    as_tibble()

  bind_cols(identity, checkpoint_info, logged, final_metrics, training_metrics, run_args)
}

# =============================================================================
# Sweep-level collection
# =============================================================================
collect <- function(sweep_dir) {
  run_dirs <- dir_ls(sweep_dir, type = "directory") |> sort()
  rows <- map(run_dirs, collect_one_run) |> compact()

  if (length(rows) == 0) return(tibble())

  result <- bind_rows(rows)

  best <- load_best_run_names(sweep_dir)
  result |> mutate(
    is_best_span     = run == best$span,
    is_best_standard = run == best$standard
  )
}

collect_fixed_eval <- function(sweep_dir, sweep_results) {
  fixed_path <- path(sweep_dir, "fixed_eval_per_prob_results.csv")
  if (!file_exists(fixed_path)) return(NULL)

  fixed <- read_csv(fixed_path, show_col_types = FALSE)

  meta <- sweep_results |>
    select(run, size, is_best_span, is_best_standard) |>
    distinct()

  left_join(fixed, meta, by = c("run_name" = "run"))
}

# =============================================================================
# CLI
# =============================================================================
parser <- ArgumentParser()
parser$add_argument(
  "--sweep",
  default = DEFAULT_SWEEP_DIR,
  help    = "Sweep directory containing run subdirectories."
)
parser$add_argument(
  "--out",
  default = NULL,
  help    = "Output CSV path for sweep results. Defaults to <sweep>/sweep_results.csv."
)
parser$add_argument(
  "--fixed-eval-out",
  default = NULL,
  help    = paste(
    "Output CSV for fixed-eval results joined with sweep metadata.",
    "Defaults to <sweep>/fixed_eval_collected.csv.",
    "Skipped if fixed_eval_per_prob_results.csv is absent."
  )
)

args      <- parser$parse_args()
sweep_dir <- args$sweep
out_file  <- args$out %||% path(sweep_dir, "sweep_results.csv")
fixed_out <- args$fixed_eval_out %||% path(sweep_dir, "fixed_eval_collected.csv")

rows <- collect(sweep_dir)
if (nrow(rows) == 0) {
  message("No results found.")
  quit(status = 1)
}
write_csv(rows, out_file)
message("Wrote ", nrow(rows), " rows to ", out_file)

fixed <- collect_fixed_eval(sweep_dir, rows)
if (!is.null(fixed)) {
  write_csv(fixed, fixed_out)
  message("Wrote ", nrow(fixed), " rows to ", fixed_out)
} else {
  message("No fixed_eval_per_prob_results.csv found — skipping fixed eval output.")
}
